pkg/storaged/btrfs/subvolume.jsx

```diff
@@ -242,7 +245,70 @@
     });
 }
 
-export function make_btrfs_subvolume_page(parent, volume, subvol) {
+function dirname(path) {
+    const i = path.lastIndexOf("/");
+    if (i < 0)
+        return null;
+    else
+        return path.substr(0, i);
+}
+
+export function make_btrfs_subvolume_pages(parent, volume) {
+    let subvols = client.uuids_btrfs_subvols[volume.data.uuid];
+    if (!subvols) {
+        const block = client.blocks[volume.path];
+        /*
+         * Try to show subvolumes based on fstab entries, this is a bit tricky
+         * as mounts where subvolid cannot be shown userfriendly.
+         *
+         * The real subvolume data structure has "id" fields and
+         * "parent" fields that refer to the ids to form a tree.  We
+         * want to do the same here, and we give fake ids to our fake
+         * subvolumes for this reason.  We don't store these fake ids
+         * in the "id" field since we don't want them to be taken
+         * seriously by the rest of the code.
+         */
+        let fake_id = 5;
+        subvols = [{ pathname: "/", id: 5, fake_id: fake_id++ }];
+        const subvols_by_pathname = { };
+        for (const config of block.Configuration) {
+            if (config[0] == "fstab") {
+                const opts = config[1].opts;
+                if (!opts)
+                    continue;
+
+                const fstab_subvol = parse_subvol_from_options(decode_filename(opts.v));
+
+                if (fstab_subvol && fstab_subvol.pathname && fstab_subvol.pathname !== "/") {
+                    fstab_subvol.fake_id = fake_id++;
+                    subvols_by_pathname[fstab_subvol.pathname] = fstab_subvol;
+                    subvols.push(fstab_subvol);
+                }
+            }
+        }
+
+        // Find parents
+        for (const pn in subvols_by_pathname) {
+            let dn = pn;
+            while (true) {
+                dn = dirname(dn);
+                if (!dn) {
+                    subvols_by_pathname[pn].parent = 5;
+                    break;
+                } else if (subvols_by_pathname[dn]) {
+                    subvols_by_pathname[pn].parent = subvols_by_pathname[dn].fake_id;
+                    break;
```

Comment:
These 3 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-20067-20240301-083916-612a49d9-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)